File: executor.cpp
/**
 * CSP Trade Executor - C++ Version
 * Executes cash-secured put orders via Interactive Brokers TWS API.
 * 
 * Build: g++ -std=c++17 -o executor executor.cpp -I/path/to/twsapi -L/path/to/twsapi -lTwsApi -lpthread
 * 
 * Requires: IB TWS API (https://interactivebrokers.github.io/)
 */

#include <iostream>
#include <string>
#include <memory>
#include <thread>
#include <chrono>
#include <mutex>
#include <condition_variable>
#include <map>
#include <vector>
#include <cmath>

// IB TWS API headers (adjust path as needed)
// #include "EWrapper.h"
// #include "EReaderOSSignal.h"
// #include "EClientSocket.h"
// #include "Contract.h"
// #include "Order.h"

// For standalone compilation demo, we'll define minimal structures
// In production, use the actual IB TWS API headers

namespace ib {
    struct Contract {
        std::string symbol;
        std::string secType = "OPT";
        std::string exchange = "SMART";
        std::string currency = "USD";
        std::string lastTradeDateOrContractMonth;
        double strike = 0.0;
        std::string right;  // "P" for put, "C" for call
    };

    struct Order {
        int orderId = 0;
        std::string action;     // "BUY" or "SELL"
        double totalQuantity = 0;
        std::string orderType;  // "LMT" or "MKT"
        double lmtPrice = 0.0;
    };

    struct TickPrice {
        double bid = 0.0;
        double ask = 0.0;
        double last = 0.0;
    };
}

/**
 * CSP Executor class for Interactive Brokers
 */
class CSPExecutor {
private:
    std::string host_;
    int port_;
    int clientId_;
    bool connected_ = false;
    int nextOrderId_ = 0;
    
    std::mutex mtx_;
    std::condition_variable cv_;
    std::map<int, ib::TickPrice> prices_;

public:
    CSPExecutor(const std::string& host = "127.0.0.1", 
                int port = 7497, 
                int clientId = 1)
        : host_(host), port_(port), clientId_(clientId) {}
    
    ~CSPExecutor() {
        disconnect();
    }

    /**
     * Connect to TWS/Gateway
     */
    bool connect() {
        std::cout << "[INFO] Connecting to IB at " << host_ << ":" << port_ << std::endl;
        
        // In production: use EClientSocket::eConnect()
        // this->client_->eConnect(host_.c_str(), port_, clientId_);
        
        // Simulated connection for demo
        connected_ = true;
        nextOrderId_ = 1000;
        
        std::cout << "[INFO] Connected successfully" << std::endl;
        return true;
    }

    /**
     * Disconnect from TWS/Gateway
     */
    void disconnect() {
        if (connected_) {
            // In production: client_->eDisconnect();
            connected_ = false;
            std::cout << "[INFO] Disconnected" << std::endl;
        }
    }

    /**
     * Create a put option contract
     */
    ib::Contract createPutContract(const std::string& symbol,
                                    double strike,
                                    const std::string& expiry) {
        ib::Contract contract;
        contract.symbol = symbol;
        contract.secType = "OPT";
        contract.exchange = "SMART";
        contract.currency = "USD";
        contract.strike = strike;
        contract.right = "P";  // Put
        
        // Format expiry: YYYYMMDD
        contract.lastTradeDateOrContractMonth = expiry;
        
        return contract;
    }

    /**
     * Request market data for an option
     */
    ib::TickPrice getOptionPrice(const ib::Contract& contract) {
        ib::TickPrice price;
        
        // In production: use reqMktData() and wait for tickPrice callbacks
        // Simulated prices for demo
        price.bid = 2.50;
        price.ask = 2.60;
        price.last = 2.55;
        
        return price;
    }

    /**
     * Sell to open a cash-secured put
     */
    struct ExecutionResult {
        std::string symbol;
        double strike;
        std::string expiry;
        int quantity;
        std::string orderType;
        double limitPrice;
        double bidPrice;
        double askPrice;
        double estimatedPremium;
        double collateralRequired;
        std::string status;
        int orderId;
        bool dryRun;
    };

    ExecutionResult sellPut(const std::string& symbol,
                            double strike,
                            const std::string& expiry,
                            int quantity = 1,
                            double limitPrice = 0.0,
                            bool dryRun = true) {
        
        ExecutionResult result;
        result.symbol = symbol;
        result.strike = strike;
        result.expiry = expiry;
        result.quantity = quantity;
        result.dryRun = dryRun;
        
        // Create contract
        auto contract = createPutContract(symbol, strike, expiry);
        
        // Get current prices
        auto prices = getOptionPrice(contract);
        result.bidPrice = prices.bid;
        result.askPrice = prices.ask;
        
        // Calculate values
        double midPrice = (prices.bid + prices.ask) / 2.0;
        result.estimatedPremium = midPrice * 100.0 * quantity;
        result.collateralRequired = strike * 100.0 * quantity;
        
        // Create order
        ib::Order order;
        order.action = "SELL";
        order.totalQuantity = quantity;
        
        if (limitPrice > 0) {
            order.orderType = "LMT";
            order.lmtPrice = limitPrice;
            result.orderType = "LIMIT";
            result.limitPrice = limitPrice;
        } else {
            order.orderType = "MKT";
            result.orderType = "MARKET";
            result.limitPrice = 0;
        }
        
        if (dryRun) {
            std::cout << "[DRY RUN] Would sell " << quantity << "x " 
                      << symbol << " $" << strike << " Put @ " 
                      << (limitPrice > 0 ? std::to_string(limitPrice) : "MKT") << std::endl;
            result.status = "DRY_RUN";
            result.orderId = 0;
        } else {
            // In production: placeOrder(contract, order)
            order.orderId = nextOrderId_++;
            result.orderId = order.orderId;
            result.status = "SUBMITTED";
            std::cout << "[LIVE] Order submitted: ID " << order.orderId << std::endl;
        }
        
        return result;
    }

    /**
     * Execute multiple trades from screener output
     */
    struct TradeInput {
        std::string ticker;
        double strike;
        std::string exp;
        double mid;
    };

    std::vector<ExecutionResult> executeFromScreener(
            const std::vector<TradeInput>& trades,
            int maxPositions = 5,
            double maxCollateral = 50000.0,
            bool dryRun = true) {
        
        std::vector<ExecutionResult> results;
        double totalCollateral = 0.0;
        
        for (size_t i = 0; i < trades.size() && (int)results.size() < maxPositions; ++i) {
            const auto& trade = trades[i];
            double collateralNeeded = trade.strike * 100.0;
            
            if (totalCollateral + collateralNeeded > maxCollateral) {
                std::cout << "[INFO] Skipping " << trade.ticker 
                          << " - would exceed max collateral" << std::endl;
                continue;
            }
            
            auto result = sellPut(
                trade.ticker,
                trade.strike,
                trade.exp,
                1,
                trade.mid,
                dryRun
            );
            
            results.push_back(result);
            totalCollateral += collateralNeeded;
        }
        
        return results;
    }
};

/**
 * Print execution result
 */
void printResult(const CSPExecutor::ExecutionResult& r) {
    std::cout << "\n📋 Trade Result:" << std::endl;
    std::cout << "   Symbol: " << r.symbol << std::endl;
    std::cout << "   Strike: $" << r.strike << std::endl;
    std::cout << "   Expiry: " << r.expiry << std::endl;
    std::cout << "   Quantity: " << r.quantity << std::endl;
    std::cout << "   Bid/Ask: $" << r.bidPrice << "/$" << r.askPrice << std::endl;
    std::cout << "   Est. Premium: $" << r.estimatedPremium << std::endl;
    std::cout << "   Collateral: $" << r.collateralRequired << std::endl;
    std::cout << "   Status: " << r.status << std::endl;
}

/**
 * Print usage
 */
void printUsage(const char* progName) {
    std::cout << "CSP Trade Executor - C++ Version\n\n"
              << "Usage:\n"
              << "  " << progName << " --symbol NVDA --strike 120 --expiry 20260321\n"
              << "  " << progName << " --demo\n\n"
              << "Options:\n"
              << "  --host      TWS/Gateway host (default: 127.0.0.1)\n"
              << "  --port      Port (default: 7497 for paper trading)\n"
              << "  --symbol    Underlying symbol\n"
              << "  --strike    Strike price\n"
              << "  --expiry    Expiration date (YYYYMMDD)\n"
              << "  --quantity  Number of contracts (default: 1)\n"
              << "  --limit     Limit price (omit for market order)\n"
              << "  --live      Actually execute (default is dry run)\n"
              << "  --demo      Run demo with sample trades\n";
}

int main(int argc, char* argv[]) {
    std::string host = "127.0.0.1";
    int port = 7497;
    int clientId = 1;
    std::string symbol;
    double strike = 0.0;
    std::string expiry;
    int quantity = 1;
    double limitPrice = 0.0;
    bool dryRun = true;
    bool demo = false;

    // Parse command line arguments
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        
        if (arg == "--host" && i + 1 < argc) host = argv[++i];
        else if (arg == "--port" && i + 1 < argc) port = std::stoi(argv[++i]);
        else if (arg == "--symbol" && i + 1 < argc) symbol = argv[++i];
        else if (arg == "--strike" && i + 1 < argc) strike = std::stod(argv[++i]);
        else if (arg == "--expiry" && i + 1 < argc) expiry = argv[++i];
        else if (arg == "--quantity" && i + 1 < argc) quantity = std::stoi(argv[++i]);
        else if (arg == "--limit" && i + 1 < argc) limitPrice = std::stod(argv[++i]);
        else if (arg == "--live") dryRun = false;
        else if (arg == "--demo") demo = true;
        else if (arg == "--help" || arg == "-h") {
            printUsage(argv[0]);
            return 0;
        }
    }

    // Create executor
    CSPExecutor executor(host, port, clientId);
    
    if (!executor.connect()) {
        std::cerr << "Failed to connect to IB. Make sure TWS/Gateway is running." << std::endl;
        return 1;
    }

    if (demo) {
        // Demo mode with sample trades
        std::cout << "\n🎯 Running demo with sample trades...\n" << std::endl;
        
        std::vector<CSPExecutor::TradeInput> sampleTrades = {
            {"NVDA", 120.0, "20260321", 3.50},
            {"AMD", 140.0, "20260321", 2.80},
            {"MSFT", 380.0, "20260321", 5.20},
            {"GOOGL", 175.0, "20260321", 4.10},
            {"META", 550.0, "20260321", 8.50}
        };
        
        auto results = executor.executeFromScreener(sampleTrades, 5, 100000.0, dryRun);
        
        std::cout << "\n📊 Execution Summary (" << results.size() << " trades):" << std::endl;
        for (const auto& r : results) {
            std::string status = (r.status == "DRY_RUN" || r.status == "SUBMITTED") ? "✅" : "⏳";
            std::cout << "   " << status << " " << r.symbol 
                      << " $" << r.strike << " Put - " << r.status << std::endl;
        }
    }
    else if (!symbol.empty() && strike > 0 && !expiry.empty()) {
        // Single trade mode
        auto result = executor.sellPut(symbol, strike, expiry, quantity, limitPrice, dryRun);
        printResult(result);
    }
    else {
        printUsage(argv[0]);
    }

    executor.disconnect();
    return 0;
}

File: executor.py
#!/usr/bin/env python3
"""
CSP Trade Executor
Executes cash-secured put orders via Interactive Brokers API.
Requires: pip install ib_insync

Usage:
    python executor.py --symbol NVDA --strike 120 --expiry 2026-03-21 --quantity 1
    python executor.py --from-json screened_trades.json
"""

import argparse
import json
import logging
from datetime import datetime
from typing import Optional

from ib_insync import IB, Option, LimitOrder, MarketOrder, util

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s [%(levelname)s] %(message)s'
)
logger = logging.getLogger(__name__)


class CSPExecutor:
    """Execute Cash-Secured Put trades via Interactive Brokers."""
    
    def __init__(self, host: str = '127.0.0.1', port: int = 7497, client_id: int = 1):
        """
        Initialize the executor.
        
        Args:
            host: TWS/Gateway host (default localhost)
            port: 7497 for TWS paper, 7496 for live, 4001/4002 for Gateway
            client_id: Unique client ID for this connection
        """
        self.ib = IB()
        self.host = host
        self.port = port
        self.client_id = client_id
        self.connected = False
    
    def connect(self) -> bool:
        """Connect to Interactive Brokers TWS/Gateway."""
        try:
            self.ib.connect(self.host, self.port, clientId=self.client_id)
            self.connected = True
            logger.info(f"Connected to IB at {self.host}:{self.port}")
            return True
        except Exception as e:
            logger.error(f"Failed to connect: {e}")
            return False
    
    def disconnect(self):
        """Disconnect from IB."""
        if self.connected:
            self.ib.disconnect()
            self.connected = False
            logger.info("Disconnected from IB")
    
    def get_account_summary(self) -> dict:
        """Get account summary including buying power."""
        summary = {}
        for item in self.ib.accountSummary():
            if item.tag in ['NetLiquidation', 'BuyingPower', 'AvailableFunds', 'TotalCashValue']:
                summary[item.tag] = float(item.value)
        return summary
    
    def create_put_contract(
        self,
        symbol: str,
        strike: float,
        expiry: str,
        exchange: str = 'SMART',
        currency: str = 'USD'
    ) -> Option:
        """
        Create an option contract for a put.
        
        Args:
            symbol: Underlying symbol (e.g., 'NVDA')
            strike: Strike price
            expiry: Expiration date as 'YYYY-MM-DD' or 'YYYYMMDD'
            exchange: Exchange (default SMART for best execution)
            currency: Currency (default USD)
        
        Returns:
            Option contract object
        """
        # Format expiry as YYYYMMDD
        if '-' in expiry:
            expiry = expiry.replace('-', '')
        
        contract = Option(
            symbol=symbol,
            lastTradeDateOrContractMonth=expiry,
            strike=strike,
            right='P',  # Put
            exchange=exchange,
            currency=currency
        )
        
        # Qualify the contract to get full details
        qualified = self.ib.qualifyContracts(contract)
        if qualified:
            return qualified[0]
        return contract
    
    def get_option_price(self, contract: Option) -> dict:
        """Get current bid/ask/mid for an option."""
        self.ib.reqMktData(contract, '', False, False)
        self.ib.sleep(2)  # Wait for data
        
        ticker = self.ib.ticker(contract)
        return {
            'bid': ticker.bid if ticker.bid > 0 else None,
            'ask': ticker.ask if ticker.ask > 0 else None,
            'mid': (ticker.bid + ticker.ask) / 2 if ticker.bid > 0 and ticker.ask > 0 else None,
            'last': ticker.last if ticker.last > 0 else None
        }
    
    def sell_put(
        self,
        symbol: str,
        strike: float,
        expiry: str,
        quantity: int = 1,
        limit_price: Optional[float] = None,
        dry_run: bool = True
    ) -> dict:
        """
        Sell to open a cash-secured put.
        
        Args:
            symbol: Underlying symbol
            strike: Strike price
            expiry: Expiration date
            quantity: Number of contracts (default 1)
            limit_price: Limit price (None for market order)
            dry_run: If True, don't actually place the order
        
        Returns:
            Order result dictionary
        """
        contract = self.create_put_contract(symbol, strike, expiry)
        
        # Get current pricing
        prices = self.get_option_price(contract)
        logger.info(f"Current prices for {symbol} ${strike} Put {expiry}: {prices}")
        
        # Create order (sell to open = negative quantity)
        if limit_price:
            order = LimitOrder('SELL', quantity, limit_price)
        else:
            order = MarketOrder('SELL', quantity)
        
        # Calculate collateral required
        collateral = strike * 100 * quantity
        premium = (prices['mid'] or prices['bid'] or 0) * 100 * quantity
        
        result = {
            'symbol': symbol,
            'strike': strike,
            'expiry': expiry,
            'quantity': quantity,
            'order_type': 'LIMIT' if limit_price else 'MARKET',
            'limit_price': limit_price,
            'current_bid': prices['bid'],
            'current_ask': prices['ask'],
            'current_mid': prices['mid'],
            'estimated_premium': premium,
            'collateral_required': collateral,
            'dry_run': dry_run
        }
        
        if dry_run:
            logger.info(f"[DRY RUN] Would sell {quantity}x {symbol} ${strike} Put @ {limit_price or 'MKT'}")
            result['status'] = 'DRY_RUN'
        else:
            trade = self.ib.placeOrder(contract, order)
            self.ib.sleep(1)
            result['status'] = trade.orderStatus.status
            result['order_id'] = trade.order.orderId
            logger.info(f"Order placed: {trade.orderStatus.status}")
        
        return result
    
    def execute_from_screener(
        self,
        trades: list,
        max_positions: int = 5,
        max_collateral: float = 50000,
        dry_run: bool = True
    ) -> list:
        """
        Execute multiple trades from screener output.
        
        Args:
            trades: List of trade dicts with 'ticker', 'strike', 'exp' keys
            max_positions: Maximum number of positions to open
            max_collateral: Maximum total collateral to use
            dry_run: If True, don't actually place orders
        
        Returns:
            List of execution results
        """
        results = []
        total_collateral = 0
        
        for i, trade in enumerate(trades[:max_positions]):
            collateral_needed = trade['strike'] * 100
            
            if total_collateral + collateral_needed > max_collateral:
                logger.info(f"Skipping {trade['ticker']} - would exceed max collateral")
                continue
            
            result = self.sell_put(
                symbol=trade['ticker'],
                strike=trade['strike'],
                expiry=trade['exp'],
                quantity=1,
                limit_price=trade.get('mid') or trade.get('bid'),
                dry_run=dry_run
            )
            
            results.append(result)
            total_collateral += collateral_needed
        
        return results


def main():
    parser = argparse.ArgumentParser(description='Execute CSP trades via Interactive Brokers')
    
    # Connection settings
    parser.add_argument('--host', default='127.0.0.1', help='TWS/Gateway host')
    parser.add_argument('--port', type=int, default=7497, help='Port (7497=TWS paper, 7496=TWS live)')
    parser.add_argument('--client-id', type=int, default=1, help='Client ID')
    
    # Single trade mode
    parser.add_argument('--symbol', '-s', help='Underlying symbol')
    parser.add_argument('--strike', '-k', type=float, help='Strike price')
    parser.add_argument('--expiry', '-e', help='Expiration date (YYYY-MM-DD)')
    parser.add_argument('--quantity', '-q', type=int, default=1, help='Number of contracts')
    parser.add_argument('--limit', '-l', type=float, help='Limit price (omit for market)')
    
    # Batch mode
    parser.add_argument('--from-json', help='Execute trades from JSON file')
    parser.add_argument('--max-positions', type=int, default=5, help='Max positions to open')
    parser.add_argument('--max-collateral', type=float, default=50000, help='Max collateral')
    
    # Safety
    parser.add_argument('--live', action='store_true', help='Actually execute (default is dry run)')
    
    args = parser.parse_args()
    dry_run = not args.live
    
    executor = CSPExecutor(host=args.host, port=args.port, client_id=args.client_id)
    
    if not executor.connect():
        print("Failed to connect to IB. Make sure TWS/Gateway is running.")
        return 1
    
    try:
        # Show account info
        summary = executor.get_account_summary()
        print(f"\n📊 Account Summary:")
        for k, v in summary.items():
            print(f"   {k}: ${v:,.2f}")
        print()
        
        if args.from_json:
            # Batch mode
            with open(args.from_json) as f:
                trades = json.load(f)
            
            results = executor.execute_from_screener(
                trades,
                max_positions=args.max_positions,
                max_collateral=args.max_collateral,
                dry_run=dry_run
            )
            
            print(f"\n📋 Execution Results ({len(results)} trades):")
            for r in results:
                status = "✅" if r['status'] in ['Filled', 'DRY_RUN'] else "⏳"
                print(f"   {status} {r['symbol']} ${r['strike']} Put - {r['status']}")
        
        elif args.symbol and args.strike and args.expiry:
            # Single trade mode
            result = executor.sell_put(
                symbol=args.symbol,
                strike=args.strike,
                expiry=args.expiry,
                quantity=args.quantity,
                limit_price=args.limit,
                dry_run=dry_run
            )
            
            print(f"\n📋 Trade Result:")
            print(f"   Symbol: {result['symbol']}")
            print(f"   Strike: ${result['strike']}")
            print(f"   Expiry: {result['expiry']}")
            print(f"   Quantity: {result['quantity']}")
            print(f"   Bid/Ask: ${result['current_bid']}/{result['current_ask']}")
            print(f"   Est. Premium: ${result['estimated_premium']:,.2f}")
            print(f"   Collateral: ${result['collateral_required']:,.2f}")
            print(f"   Status: {result['status']}")
        
        else:
            parser.print_help()
    
    finally:
        executor.disconnect()
    
    return 0


if __name__ == '__main__':
    exit(main())
